From 74250a7cdf106d4ca7d9506e6d5dc7c448dc3434 Mon Sep 17 00:00:00 2001
From: David Michael <david.michael@coreos.com>
Date: Thu, 15 Dec 2016 15:22:57 -0800
Subject: [PATCH] gshadow: Sync fgetsgent_r.c with grp/fgetgrent_r.c

	[BZ #20338]
	* gshadow/fgetsgent_r.c: Include <libio/iolibio.h>.
	(flockfile): New macro.
	(funlockfile): Likewise.
	(__fgetsgent_r): Sync with __fgetgrent_r.
	* nss/nss_files/files-sgrp.c: Fix "fgetsgent_r.c" typo.
---

Flatcar Container Linux notes: the patch is a back-port from glibc-2.32
to fix an issue where parsing of /etc/gshadow can return bad pointers
causing segfaults in applications. Details e.g. here:
https://bugs.launchpad.net/ubuntu/+source/glibc/+bug/1890535

Forward-ported from glibc-2.25 to 2.30 by the Flatcar Container Linux
team.

diff --git a/gshadow/fgetsgent_r.c b/gshadow/fgetsgent_r.c
index 19a663f0..81c56125 100644
--- a/gshadow/fgetsgent_r.c
+++ b/gshadow/fgetsgent_r.c
@@ -20,6 +20,10 @@
 #include <gshadow.h>
 #include <stdio.h>
 
+#include <libio/iolibio.h>
+#define flockfile(s) _IO_flockfile (s)
+#define funlockfile(s) _IO_funlockfile (s)
+
 /* Define a line parsing function using the common code
    used in the nss_files module.  */
 
@@ -31,28 +35,29 @@ struct sgent_data {};
 #include <nss/nss_files/files-parse.c>
 
 
-/* Read one shadow entry from the given stream.  */
+/* Read one entry from the given stream.  */
 int
 __fgetsgent_r (FILE *stream, struct sgrp *resbuf, char *buffer, size_t buflen,
 	       struct sgrp **result)
 {
   char *p;
+  int parse_result;
 
-  _IO_flockfile (stream);
+  flockfile (stream);
   do
     {
       buffer[buflen - 1] = '\xff';
       p = fgets_unlocked (buffer, buflen, stream);
-      if (p == NULL && feof_unlocked (stream))
+      if (__builtin_expect (p == NULL, 0) && feof_unlocked (stream))
 	{
-	  _IO_funlockfile (stream);
+	  funlockfile (stream);
 	  *result = NULL;
 	  __set_errno (ENOENT);
 	  return errno;
 	}
-      if (p == NULL || buffer[buflen - 1] != '\xff')
+      if (__builtin_expect (p == NULL, 0) || buffer[buflen - 1] != '\xff')
 	{
-	  _IO_funlockfile (stream);
+	  funlockfile (stream);
 	  *result = NULL;
 	  __set_errno (ERANGE);
 	  return errno;
@@ -64,10 +69,17 @@ __fgetsgent_r (FILE *stream, struct sgrp *resbuf, char *buffer, size_t buflen,
     } while (*p == '\0' || *p == '#' /* Ignore empty and comment lines.  */
 	     /* Parse the line.  If it is invalid, loop to
 		get the next line of the file to parse.  */
-	     || ! parse_line (buffer, (void *) resbuf, (void *) buffer, buflen,
-			      &errno));
+	     || ! (parse_result = parse_line (p, resbuf,
+					      (void *) buffer, buflen,
+					      &errno)));
+  funlockfile (stream);
 
-  _IO_funlockfile (stream);
+  if (__builtin_expect (parse_result, 0) == -1)
+    {
+      /* The parser ran out of space.  */
+      *result = NULL;
+      return errno;
+    }
 
   *result = resbuf;
   return 0;

diff --git a/nss/nss_files/files-sgrp.c b/nss/nss_files/files-sgrp.c
index 15dc659..05c3805 100644
--- a/nss/nss_files/files-sgrp.c
+++ b/nss/nss_files/files-sgrp.c
@@ -23,7 +23,7 @@
 #define DATABASE	"gshadow"
 struct sgent_data {};
 
-/* Our parser function is already defined in sgetspent_r.c, so use that
+/* Our parser function is already defined in sgetsgent_r.c, so use that
    to parse lines from the database file.  */
 #define EXTERN_PARSER
 #include "files-parse.c"
-- 
2.7.4
